jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceDecompositionBuilder.java

```diff
@@ -0,0 +1,210 @@
+ * forget node, the corresponding bag of the introduce node is smaller than the set of neighbors of
+ * this vertex. Thus the time complexity is in $\mathcal{O}(|V|(|V|+|E|))$.
+ * <p>
+ * This is a non-recursive adaption for nice tree decomposition of algorithm 2 from here: <br>
+ * Hans L. Bodlaender, Arie M.C.A. Koster, Treewidth computations I. Upper bounds, Information and
+ * Computation, Volume 208, Issue 3, 2010, Pages 259-275,
+ * 
+ * @author Ira Justus Fesefeldt (PhoenixIra)
+ * @author Timofey Chudakov
+ *
+ * @since June 2018
+ *
+ * @param <V> the vertex type of the graph
+ * @param <E> the edge type of the graph
+ */
+public class ChordalNiceDecompositionBuilder<V, E>
+    extends
+    NiceDecompositionBuilder<V>
+{
+    // the chordal graph
+    private Graph<V, E> graph;
+
+    // the perfect elimination order of graph
+    private List<V> perfectOrder;
+
+    // another representation of the perfect elimination order of graph
+    private Map<V, Integer> vertexInOrder;
+
+    /**
+     * Constructor for the nice decomposition builder of chordal graphs. The constructor
+     * already computes the decomposition.
+     * 
+     * @throws IllegalArgumentException if the graph is not chordal.
+     * @param graph the chordal graph for which a decomposition should be created
+     */
+    public ChordalNiceDecompositionBuilder(Graph<V, E> graph)
+    {
+        super();
+        ChordalityInspector<V, E> inspec = new ChordalityInspector<>(graph);
+        if (!inspec.isChordal())
+            throw new IllegalArgumentException("The given graph is not chordal.");
+        this.graph = graph;
+        this.perfectOrder = inspec.getPerfectEliminationOrder();
+        vertexInOrder = getVertexInOrder();
+        computeNiceDecomposition();
+    }
+
+    /**
+     * Constructor for the nice decomposition builder of chordal graphs. 
+     * This method needs the perfect elimination order. The constructor
+     * already computes the decomposition.<p>
+     * Note: This method does NOT check whether the order is correct.
+     * 
+     * @param graph the chordal graph for which a decomposition should be created
+     * @param perfectEliminationOrder the perfect elimination order of the graph
+     */
+    public ChordalNiceDecompositionBuilder(Graph<V, E> graph, List<V> perfectEliminationOrder)
+    {
+        super();
+        this.graph = graph;
+        this.perfectOrder = perfectEliminationOrder;
+        computeNiceDecomposition();
+    }
+
+    /**
+     * Computes the nice decomposition of the graph. We iterate over the perfect elimination order
+     * of the chordal graph and try to add a node containing the predecessors regarding the
+     * perfect elimination as a bag to the tree
+     */
+    private void computeNiceDecomposition()
+    {
+        // map from vertices to decomposition-nodes where the decomposition-node has all its
+        // predecessors
```

Comment:
Let's mention that here the decomposition nodes are forget nodes and they also contain corresponding vertices:
```
/*
 * Map from the vertices of the graph to the forget nodes from the decomposition tree. The mapping 
 * is one-to-one and each decomposition node contains corresponding vertex and all its predecessors.
 */
```